enterprise/server/soci/soci.go

```diff
@@ -0,0 +1,316 @@
+package soci
+
+import (
+	"context"
+	"encoding/json"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/buildbuddy-io/buildbuddy/enterprise/server/remote_execution/commandutil"
+	repb "github.com/buildbuddy-io/buildbuddy/proto/remote_execution"
+	rspb "github.com/buildbuddy-io/buildbuddy/proto/resource"
+	socipb "github.com/buildbuddy-io/buildbuddy/proto/soci"
+	"github.com/buildbuddy-io/buildbuddy/server/environment"
+	"github.com/buildbuddy-io/buildbuddy/server/interfaces"
+	"github.com/buildbuddy-io/buildbuddy/server/util/log"
+	"github.com/buildbuddy-io/buildbuddy/server/util/prefix"
+	"github.com/buildbuddy-io/buildbuddy/server/util/status"
+)
+
+type SociArtifactStore struct {
+	cache     interfaces.Cache
+	blobstore interfaces.Blobstore
+	env       environment.Env
+}
+
+func Register(env environment.Env) error {
+	err, server := NewSociArtifactStore(env)
+	if err != nil {
+		return err
+	}
+	env.SetSociArtifactStoreServer(server)
+	return nil
+}
+
+func NewSociArtifactStore(env environment.Env) (error, *SociArtifactStore) {
+	if env.GetCache() == nil {
+		return status.InternalError("soci artifact server requires a cache"), nil
+	}
+	if env.GetBlobstore() == nil {
+		return status.InternalError("soci artifact server requires a blobstore"), nil
+	}
+	return nil, &SociArtifactStore{
+		cache:     env.GetCache(),
+		blobstore: env.GetBlobstore(),
+		env:       env,
+	}
+}
+
+func (s *SociArtifactStore) GetArtifacts(ctx context.Context, req *socipb.GetArtifactsRequest) (*socipb.GetArtifactsResponse, error) {
+	ctx, err := prefix.AttachUserPrefixToContext(ctx, s.env)
+	if err != nil {
+		return nil, err
+	}
+	imageId, err := imageId(ctx, req.ImageName)
+	if err != nil {
+		return nil, err
+	}
+	exists, err := s.blobstore.BlobExists(ctx, blobKey(imageId))
+	if err != nil {
+		return nil, err
+	}
+	if exists {
+		return s.getArtifactsFromCache(ctx, imageId)
+	}
+	// TODO(iain): add a mutex to prevent multiple parallel calls
+	sociIndexDigest, ztocDigests, err := pullAndIndexImage(ctx, req.ImageName, imageId)
+	if err != nil {
+		return nil, err
+	}
+	if err = s.writeArtifactsToCache(ctx, imageId, sociIndexDigest, ztocDigests); err != nil {
+		return nil, err
+	}
+	return getArtifactsResponse(imageId, sociIndexDigest, ztocDigests), nil
+}
+
+type ImageInfo struct {
+	Digest string `json:"Digest"`
+}
+
+// The image id is the digest of the actual image and changes as image tags are
+// altered. Grab it by running `skopeo inspect`.
+func imageId(ctx context.Context, imageName string) (string, error) {
+	cmd := []string{"skopeo", "inspect", "docker://" + imageName}
```

Comment:
Do we need `skopeo` in the mix? Container Registry APIs are relatively simple. Take a look at `registry.go` which has code for getting image information.

Reply:
Nope, looks like `registry.go` does the trick too, thanks for the recommendation.